File: src/extras-kcoreaddons/libkjsonutils.cpp
#include <KJsonUtils>
#include <QJsonObject>
#include <QJsonValue>
#include <QString>
#include <kjsonutils.h>
#include "libkjsonutils.h"
#include "libkjsonutils.hxx"

QJsonValue* KJsonUtils_ReadTranslatedValue(const QJsonObject* jo, const libqt_string key, const QJsonValue* defaultValue) {
    QString key_QString = QString::fromUtf8(key.data, key.len);
    return new QJsonValue(KJsonUtils::readTranslatedValue(*jo, key_QString, *defaultValue));
}

libqt_string KJsonUtils_ReadTranslatedString(const QJsonObject* jo, const libqt_string key, const libqt_string defaultValue) {
    QString key_QString = QString::fromUtf8(key.data, key.len);
    QString defaultValue_QString = QString::fromUtf8(defaultValue.data, defaultValue.len);
    auto _ret = KJsonUtils::readTranslatedString(*jo, key_QString, defaultValue_QString);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

File: src/extras-kcoreaddons/libkjsonutils.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const QJsonObject = @import("libqt6").QJsonObject;
const QJsonValue = @import("libqt6").QJsonValue;
const std = @import("std");

/// ### [Upstream resources](https://api.kde.org/kjsonutils.html)
pub const KJsonUtils = extern struct {
    /// ### [Upstream resources](https://api.kde.org/kjsonutils.html)
    ///
    /// The pointer to the underlying Qt C++ object
    ///
    ptr: QtC.KJsonUtils,

    pub const _is_KJsonUtils = {};

    /// ### [Upstream resources](https://api.kde.org/kjsonutils.html#readTranslatedValue)
    ///
    /// ## Parameter(s):
    ///
    /// ` jo: QJsonObject `
    ///
    /// ` key: []const u8 `
    ///
    /// ` defaultValue: QJsonValue `
    ///
    pub fn ReadTranslatedValue(jo: anytype, key: []const u8, defaultValue: anytype) QJsonValue {
        comptime _ = @TypeOf(jo)._is_QJsonObject;
        const key_str = qtc.libqt_string{
            .len = key.len,
            .data = key.ptr,
        };
        comptime _ = @TypeOf(defaultValue)._is_QJsonValue;
        return .{ .ptr = qtc.KJsonUtils_ReadTranslatedValue(@ptrCast(jo.ptr), key_str, @ptrCast(defaultValue.ptr)) };
    }

    /// ### [Upstream resources](https://api.kde.org/kjsonutils.html#readTranslatedString)
    ///
    /// ## Parameter(s):
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    /// ` jo: QJsonObject `
    ///
    /// ` key: []const u8 `
    ///
    /// ` defaultValue: []const u8 `
    ///
    pub fn ReadTranslatedString(allocator: std.mem.Allocator, jo: anytype, key: []const u8, defaultValue: []const u8) []const u8 {
        comptime _ = @TypeOf(jo)._is_QJsonObject;
        const key_str = qtc.libqt_string{
            .len = key.len,
            .data = key.ptr,
        };
        const defaultValue_str = qtc.libqt_string{
            .len = defaultValue.len,
            .data = defaultValue.ptr,
        };
        var _str = qtc.KJsonUtils_ReadTranslatedString(@ptrCast(jo.ptr), key_str, defaultValue_str);
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kjsonutils.ReadTranslatedString: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }
};
